### PR TITLE
Add retry strategy for publishing

### PR DESCRIPTION
**Pull request by bentsherman**
This PR does the following:

- Adds retry logic for publish errors
- Change `failOnError` to be true by default

For now, the retry params are hard-coded and all exceptions are retried.

### CODE REVIEW

**Comment by pditommaso** [comment]
Nice Ben, maybe it's time to add a retry config as done for Azure and Google eg. 

https://github.com/nextflow-io/nextflow/blob/main/plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzRetryConfig.groovy#L34-L52

**Comment by ewels** [comment]
Broken link, but I guess it was meant to be this?

https://github.com/nextflow-io/nextflow/blob/744fff2a7753b75b0125caf073ad4b30a89433ef/plugins/nf-azure/src/main/nextflow/cloud/azure/config/AzRetryConfig.groovy#L33-L52

File: modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy [hunk 1]
```diff
@@ -214,6 +223,9 @@ class PublishDir {
         if( params.storageClass )
             result.storageClass = params.storageClass as String
 
+        if( params.retryPolicy )
+            result.retryConfig = new RetryConfig(params.retryPolicy)
+
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy:228]: Likely as a user, I want to configure this once for all. I'd put something in the config as `nextflow.publish.retryPolicy`

File: modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy [hunk 2]
```diff
@@ -60,7 +66,7 @@ class PublishDir {
 
     private Map<Path,Boolean> makeCache = new HashMap<>()
 
-    private Session session = Global.session as Session
+    private static Session getSession() { Global.getSession() as Session }
```
[the remark this answers is not in the document]
Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy:69]: What do you propose instead? This pattern is already used in several other places, so likely not making anything worse. I would love to remove the pattern entirely from the codebase but that is a separate effort

[the remark this answers is not in the document]
Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy:69]: Why do you need the session static?

[the remark this answers is not in the document]
Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy:69]: Nevermind, found another way

**Comment by pditommaso** [comment]
IntelliJ reports an error at [this line](https://github.com/nextflow-io/nextflow/blob/retry-publish-errors/modules/nextflow/src/main/groovy/nextflow/processor/PublishDir.groovy#L403-L403) saying the call is ambiguous

**Comment by bentsherman** [comment]
Based on how failsafe is used elsewhere in the code, my guess is the lambda must be cast to a `CheckedSupplier`, let me know if that fixes it

**Pull request merged**